modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java

```diff
@@ -285,12 +314,30 @@ private CompletableFuture<Long> queryDataNodesCount(Collection<Peer> peers, HasD
                         .thenApply(response -> {
                             assert response instanceof HasDataResponse : response;
 
-                            return ((HasDataResponse) response).result();
+                            return ((HasDataResponse) response).presence();
                         })
-                        .exceptionally(unused -> false))
+                        .exceptionally(unused -> DataPresence.UNKNOWN))
                 .toArray(CompletableFuture[]::new);
 
-        return allOf(requestFutures)
-                .thenApply(unused -> Arrays.stream(requestFutures).filter(CompletableFuture::join).count());
+        return allOf(presenceFutures)
+                .thenApply(unused -> {
+                    List<DataPresence> hasDataFlags = Arrays.stream(presenceFutures)
+                            .map(CompletableFuture::join)
+                            .collect(toList());
+
+                    long nodesSurelyHavingData = hasDataFlags.stream().filter(presence -> presence == DataPresence.HAS_DATA).count();
+                    long nodesSurelyEmpty = hasDataFlags.stream().filter(presence -> presence == DataPresence.EMPTY).count();
+                    return new DataNodesCounts(nodesSurelyHavingData, nodesSurelyEmpty);
+                });
+    }
+
+    private static class DataNodesCounts {
+        private final long nodesSurelyHavingData;
```

Comment:
These fields' names are a bit of a mouthful. Can we rename them to something like "dataNodesCnt" and "emptyNodesCnt"?

Reply:
Is it a problem? As a result, the code that uses them is self-explanatory. On the other hand, `dataNodesCnt` does not tell you that we actually checked that there is the given count of replicas *available*.

Reply:
For me, actually, the current naming was very confusing for some reason. I had to re-read them multiple times. If you don't like my suggestion, maybe you could provide something simpler?